core/Services/Calendar/Store.vala

```diff
@@ -1,307 +1,658 @@
 
-        if (event.has_recurrences () && mod_type != ECal.ObjModType.ALL) {
-            rid = event.get_recurid_as_string ();
-            debug (@"Removing recurrent event '$rid'");
+                    lock (components_add_transaction) {
+                        components_add_transaction.remove (removed_view);
+                    }
+                    removed_view.stop ();
+                    break;
+                }
+            }
         }
+    }
 
-        debug (@"Removing event '$uid'");
-        ECal.Client client;
-        lock (source_client) {
-            client = source_client.get (source.get_uid ());
+    //--- Public ECal.Component API ---//
+
+    public bool component_is_created (ECal.Component component) {
+        if (component == null) {
+            return false;
         }
+#if E_CAL_2_0
+        var created = component.get_created ();
+        return created.is_valid_time ();
+#else
+        ICal.Time created;
+        component.get_created (out created);
+        return !created.is_null_time ();
+#endif
+    }
+
+    public bool component_is_completed (ECal.Component component) {
+        return component.get_icalcomponent ().get_status () == ICal.PropertyStatus.COMPLETED;
+    }
+
+    public void component_set_status (ECal.Component component, ICal.PropertyStatus status) {
+        unowned ICal.Component ical_component = component.get_icalcomponent ();
+        ical_component.set_status (status);
+
+        switch (status) {
+            case ICal.PropertyStatus.NONE:
 
+                component.set_percent_complete (0);
 #if E_CAL_2_0
-        client.remove_object.begin (uid, rid, mod_type, ECal.OperationFlags.NONE, null, (obj, results) => {
+                component.set_completed (new ICal.Time.null_time ());
 #else
-        client.remove_object.begin (uid, rid, mod_type, null, (obj, results) => {
+                var null_time = ICal.Time.null_time ();
+                component.set_completed (ref null_time);
 #endif
-            try {
-                client.remove_object.end (results);
-            } catch (Error e) {
-                warning (e.message);
+                break;
+
+            case ICal.PropertyStatus.COMPLETED:
+                component.set_percent_complete (100);
+#if E_CAL_2_0
+                component.set_completed (new ICal.Time.today ());
+#else
+                var today_time = ICal.Time.today ();
+                component.set_completed (ref today_time);
+#endif
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    public void component_add (E.Source source, ECal.Component component) {
+        var views = source_get_views (source);
+
+        var components = new Gee.ArrayList<ECal.Component> ((Gee.EqualDataFunc<ECal.Component>?) Calendar.Util.ecalcomponent_equal_func);  // vala-lint=line-length
+        components.add (component);
+        components_added (components.read_only_view, source, views.read_only_view);
```

Comment:
The way this is designed, callbacks are called twice for every new event. This is called here, then the component is added to `transactional_components`. Then, `source_component_add` adds the component to a `Source`, which then sends the `source_icalcomponents_added` signal. That captures `transactional_components` again and calls `components_modified` for the same events.

This actually raises two questions for me. 
1. What is the purpose of `components_added_transaction`? As far as I can tell, all that happens to a component at the moment is that it's added, and then immediately removed by `source_icalcomponents_added` or a catch block in `component_add`. I don't see it used anywhere in between. And `source_icalcomponents_added` already has a list of the relevant components.
2. Why do you need to check for modified components in the component_added callback at all? If things are functioning correctly on the libecal side, the function should only be called for truly new components, unless I'm missing something. I imagine this could also answer question 1, but I'm not sure what that answer is is.

Reply:
@mcclurgm this is actually needed to optimistically show changes in the UI - even though they are not confirmed by the EDS backend yet:

- When the user adds a new event, Calendar.Store immediately fires components_added to let the UI display the new event
- Then, Calendar.Store sends the new event to the EDS backend
- Once EDS backend has added the event, we need to update the event in the UI again to avoid update errors of the same event due to internally changed E-Tags

Because EDS fires a "components_added" event itself, Calendar.Store needs a way to figure out if the event in question was already added in the UI before (and therefore needs to be updated) - or if the event was never rendered before and needs to be added. That's where `components_add_transaction` comes into play: It holds all events, which were added to the UI before and only needs updating.

In case you add an event on another device for example, the EDS "components_added" event will be fired too. Because `components_add_transaction` does not contain the event in question, Calendar.Store knows it needs to add the event to the UI instead of updating it.